File: ops/coder/figures/r/computor_figures.R
## Publish figures to the Computor figure folder from R.
##
## Workspaces are containers without a desktop, so R's interactive devices
## (x11, quartz) have nowhere to draw and a student's plot(x, y) would produce
## nothing they can see. This makes the default device an offscreen one that
## mirrors itself into the folder described in docs/figures.md, where the
## Computor VS Code extension picks it up.
##
## Enable it for every session by sourcing this file from Rprofile.site. The
## student configures nothing: a plain plot(x, y) publishes, dev.off() closes.
##
## Without COMPUTOR_FIGURES_DIR nothing here activates, so the same image can
## run gradings without leaving figure files behind.

.computor_figures <- new.env(parent = emptyenv())
.computor_figures$tracked <- list()   # device number (as character) -> entry

#' The figure folder, or NULL when publishing is switched off.
computor_figures_dir <- function() {
  configured <- trimws(Sys.getenv("COMPUTOR_FIGURES_DIR", unset = ""))
  if (nzchar(configured)) configured else NULL
}

.computor_stem <- function(number) sprintf("fig-%06d", number)

.computor_path <- function(folder, number, extension) {
  file.path(folder, paste0(.computor_stem(number), extension))
}

.computor_next_number <- function() {
  taken <- vapply(.computor_figures$tracked, function(entry) entry$number, integer(1))
  number <- 1L
  while (number %in% taken) number <- number + 1L
  number
}

#' Name a figure. Without this it is labelled "Figure N", because R carries no
#' title on the device itself and the `main=` of a plot is not readable back.
computor_figure_title <- function(title, which = grDevices::dev.cur()) {
  key <- as.character(which)
  entry <- .computor_figures$tracked[[key]]
  if (!is.null(entry)) {
    entry$title <- as.character(title)[1]
    entry$published <- FALSE   # force a rewrite of the sidecar
    .computor_figures$tracked[[key]] <- entry
  }
  invisible(title)
}

#' The device R opens when a plot needs one. An offscreen PNG with its display
#' list kept, so the page can be re-rendered to a file whenever it changes.
computor_figures_device <- function(width = 7, height = 5, res = 110, ...) {
  opened <- FALSE
  for (type in c("cairo", NA_character_)) {
    opened <- tryCatch({
      if (is.na(type)) {
        grDevices::png(tempfile(fileext = ".png"), width = width, height = height,
                       units = "in", res = res, ...)
      } else {
        grDevices::png(tempfile(fileext = ".png"), width = width, height = height,
                       units = "in", res = res, type = type, ...)
      }
      TRUE
    }, error = function(e) FALSE)
    if (opened) break
  }
  if (!opened) {
    stop("computor_figures: no usable PNG device (is cairo available?)")
  }

  grDevices::dev.control("enable")

  # Only devices opened here are published. A student's own png("out.png") is
  # their business and must not be hijacked.
  device <- grDevices::dev.cur()
  .computor_figures$tracked[[as.character(device)]] <- list(
    number = .computor_next_number(), title = NULL, published = FALSE,
    width = width, height = height, res = res
  )
  invisible(device)
}

.computor_withdraw <- function(folder, number) {
  unlink(.computor_path(folder, number, ".png"))
  unlink(.computor_path(folder, number, ".json"))
}

.computor_write_atomically <- function(folder, number, extension, write) {
  ## The viewer reacts to files appearing, so it must never get to read a
  ## half-written PNG. A rename inside the folder is atomic. The temporary name
  ## starts with a dot so it is not mistaken for a figure -- readers match
  ## fig-NNNNNN.png exactly.
  temporary <- file.path(folder, paste0(".", .computor_stem(number), ".tmp", extension))
  on.exit(unlink(temporary), add = TRUE)
  write(temporary)
  file.rename(temporary, .computor_path(folder, number, extension))
}

.computor_read_raw <- function(path) {
  if (!file.exists(path)) return(NULL)
  readBin(path, "raw", n = file.size(path))
}

.computor_publish <- function(folder, device, entry) {
  grDevices::dev.set(device)

  rendered <- tempfile(fileext = ".png")
  on.exit(unlink(rendered), add = TRUE)
  grDevices::dev.copy(grDevices::png, filename = rendered, width = entry$width,
                      height = entry$height, units = "in", res = entry$res)
  grDevices::dev.off()   # closes the copy, not the figure

  bytes <- .computor_read_raw(rendered)
  if (is.null(bytes)) return(entry)

  ## Every top-level expression triggers a sync, but most of them change
  ## nothing. Re-publishing an identical image would tell the viewer the figure
  ## updated and make it reload on every keystroke, so compare first.
  if (entry$published &&
      identical(bytes, .computor_read_raw(.computor_path(folder, entry$number, ".png")))) {
    return(entry)
  }

  title <- if (is.null(entry$title)) paste("Figure", entry$number) else entry$title
  metadata <- sprintf('{"number":%d,"title":%s,"source":"r"}',
                      entry$number, .computor_json_string(title))

  ## Metadata first: the viewer keys off the PNG, so by the time it sees the
  ## image the sidecar it reads is already in place.
  .computor_write_atomically(folder, entry$number, ".json",
                             function(path) writeLines(metadata, path, useBytes = TRUE))
  .computor_write_atomically(folder, entry$number, ".png",
                             function(path) writeBin(bytes, path))
  entry$published <- TRUE
  entry
}

.computor_json_string <- function(text) {
  escaped <- gsub("\\", "\\\\", text, fixed = TRUE)
  escaped <- gsub('"', '\\"', escaped, fixed = TRUE)
  escaped <- gsub("\n", "\\n", escaped, fixed = TRUE)
  escaped <- gsub("\r", "\\r", escaped, fixed = TRUE)
  escaped <- gsub("\t", "\\t", escaped, fixed = TRUE)
  paste0('"', escaped, '"')
}

#' Bring the folder in step with this session's devices.
#'
#' `closing` is FALSE while R is shutting down: R closes every device on the
#' way out, and treating that as "the student closed their plots" would delete
#' the figures of a script the instant it finished -- the opposite of what the
#' folder is for.
computor_figures_sync <- function(closing = TRUE) {
  folder <- computor_figures_dir()
  if (is.null(folder)) return(invisible(FALSE))
  if (!dir.exists(folder)) dir.create(folder, recursive = TRUE, showWarnings = FALSE)

  open <- grDevices::dev.list()
  keys <- if (is.null(open)) character(0) else as.character(open)
  previous <- grDevices::dev.cur()

  if (closing) {
    for (key in setdiff(names(.computor_figures$tracked), keys)) {
      .computor_withdraw(folder, .computor_figures$tracked[[key]]$number)
      .computor_figures$tracked[[key]] <- NULL
    }
  }

  for (key in intersect(names(.computor_figures$tracked), keys)) {
    entry <- .computor_figures$tracked[[key]]
    device <- as.integer(key)

    ## The viewer's close button deletes the PNG. That is the whole protocol
    ## for closing a figure from outside.
    if (closing && entry$published &&
        !file.exists(.computor_path(folder, entry$number, ".png"))) {
      .computor_withdraw(folder, entry$number)
      .computor_figures$tracked[[key]] <- NULL
      try(grDevices::dev.off(device), silent = TRUE)
      next
    }

    .computor_figures$tracked[[key]] <- tryCatch(
      .computor_publish(folder, device, entry),
      error = function(e) {
        message("computor_figures: cannot publish figure ", entry$number, ": ", conditionMessage(e))
        entry
      }
    )
  }

  ## Never restore the null device (1): dev.set(1) does not select "no device",
  ## it opens a fresh one through options(device=) -- which is this very
  ## function, so closing the last figure would silently spawn an empty one.
  if (previous != 1L && previous %in% grDevices::dev.list()) {
    try(grDevices::dev.set(previous), silent = TRUE)
  }
  invisible(TRUE)
}

#' Start publishing this session's figures. Safe to call twice.
computor_figures_start <- function() {
  if (is.null(computor_figures_dir())) return(invisible(FALSE))

  options(device = computor_figures_device)

  if (is.null(.computor_figures$callback)) {
    ## Runs after every top-level expression -- the moment a plot is finished
    ## and the device still exists. This is what makes both a script and an
    ## interactive session publish without the student asking.
    .computor_figures$callback <- addTaskCallback(
      function(expr, value, ok, visible) {
        try(computor_figures_sync(), silent = TRUE)
        TRUE
      },
      name = "computor_figures"
    )
  }
  invisible(TRUE)
}

#' Stop publishing. Figures already in the folder stay: a later session simply
#' overwrites the files it reuses.
computor_figures_stop <- function() {
  if (!is.null(.computor_figures$callback)) {
    removeTaskCallback("computor_figures")
    .computor_figures$callback <- NULL
  }
  invisible(TRUE)
}

computor_figures_start()

File: ops/coder/figures/octave/computor_figures.m
function varargout = computor_figures(action, ~)
%COMPUTOR_FIGURES Publish this session's figures to the Computor figure folder.
%   Workspaces are containers without a desktop, so an Octave figure window has
%   nowhere to appear. This mirrors every figure into the folder described in
%   docs/figures.md, where the Computor VS Code extension shows it.
%
%   computor_figures("start")  begin publishing, and keep publishing while the
%                              session sits at the prompt
%   computor_figures("sync")   bring the folder in step with the figures right
%                              now -- the whole job in one call
%   computor_figures("stop")   stop publishing (files already written stay)
%
%   Octave gives no hook that runs while a *script's* figures still hold their
%   content: atexit and onCleanup both run after the graphics are torn down,
%   plotting never routes through the m-file drawnow, and newplot wipes any
%   listener put on an axes. So an interactive session publishes by itself, via
%   add_input_event_hook, while a script needs a "sync" after it -- which the
%   octave wrapper in this folder does, so a student still configures nothing.

  persistent tracked = struct("number", {}, "figure", {}, "fingerprint", {});

  if nargin < 1 || ~ischar(action)
    action = "sync";   % add_input_event_hook calls us with its own argument
  end

  switch lower(action)
    case "start"
      folder = figures_folder();
      if isempty(folder)
        varargout{1} = false;
        return;
      end
      if ~isfolder(folder)
        mkdir(folder);
      end
      computor_figures("stop");
      add_input_event_hook("computor_figures");
      varargout{1} = true;
      return;

    case "stop"
      try
        remove_input_event_hook("computor_figures");
      catch
        % Not registered. Nothing to take back.
      end
      varargout{1} = true;
      return;

    case "sync"
      tracked = sync(tracked);
      if nargout > 0
        varargout{1} = numel(tracked);
      end
      return;

    otherwise
      error("computor_figures:action", "Unknown action '%s'", action);
  end
end


function folder = figures_folder()
%FIGURES_FOLDER The figure folder, or "" when publishing is switched off.
  folder = strtrim(getenv("COMPUTOR_FIGURES_DIR"));
end


function tracked = sync(tracked)
%SYNC Publish what changed, drop what closed, honour what the viewer removed.
  folder = figures_folder();
  if isempty(folder)
    return;
  end
  if ~isfolder(folder)
    mkdir(folder);
  end

  open = get(0, "children");            % handle-visible figures only
  open = open(:)';

  % Figures closed in Octave take their files with them.
  keep = true(1, numel(tracked));
  for k = 1:numel(tracked)
    if ~any(open == tracked(k).figure)
      withdraw(folder, tracked(k).number);
      keep(k) = false;
    end
  end
  tracked = tracked(keep);

  keep = true(1, numel(tracked));
  for k = 1:numel(open)
    handle = open(k);
    index = find([tracked.figure] == handle, 1);

    if isempty(index)
      entry = struct("number", allocate_number(tracked, handle), ...
                     "figure", handle, "fingerprint", "");
      tracked(end + 1) = entry;
      index = numel(tracked);
      keep(end + 1) = true;
    end

    entry = tracked(index);

    % The viewer's close button deletes the PNG. That is the whole protocol for
    % closing a figure from outside, so act on it before republishing -- which
    % would otherwise resurrect the file the student just closed.
    if ~isempty(entry.fingerprint) && ...
        ~isfile(figure_path(folder, entry.number, ".png"))
      withdraw(folder, entry.number);
      close(entry.figure);
      keep(index) = false;
      continue;
    end

    current = fingerprint(handle);
    if strcmp(current, entry.fingerprint)
      continue;   % nothing a student could see has changed
    end

    if publish(folder, entry.number, handle)
      entry.fingerprint = current;
      tracked(index) = entry;
    end
  end

  tracked = tracked(keep);
end


function number = allocate_number(tracked, handle)
%ALLOCATE_NUMBER The file number for a figure: its own, or the lowest free one.
%   Octave figure handles are the numbers a student sees, so figure(3) and
%   fig-000003.png stay the same thing to them.
  taken = [];
  if ~isempty(tracked)
    taken = [tracked.number];
  end

  number = handle;
  if isscalar(number) && number >= 1 && number == fix(number) && ~any(taken == number)
    return;
  end

  number = 1;
  while any(taken == number)
    number = number + 1;
  end
end


function text = fingerprint(handle)
%FINGERPRINT A cheap stand-in for "has this figure changed?".
%   Octave offers no change event that survives a plot: newplot resets the axes
%   and takes any listener with it. Rendering every figure on every tick would
%   cost a full print each time, so the visible state is summarised instead --
%   which objects exist, and what the axes show.
  parts = {sprintf("%d", numel(findall(handle))), flatten(get(handle, "name"))};

  allAxes = findall(handle, "type", "axes");
  for k = 1:numel(allAxes)
    axesHandle = allAxes(k);
    parts{end + 1} = sprintf("%.17g,", ...
        [get(axesHandle, "xlim"), get(axesHandle, "ylim"), get(axesHandle, "zlim")]);
    parts{end + 1} = sprintf("%s|%s", get(axesHandle, "xscale"), get(axesHandle, "yscale"));
    kids = get(axesHandle, "children");
    parts{end + 1} = sprintf("%.17g|", kids(:)');
  end

  % Every piece of text the student can see. The title, the axis labels and a
  % legend are text objects that already exist before they say anything, so
  % counting objects would miss xlabel("time") entirely.
  texts = findall(handle, "type", "text");
  for k = 1:numel(texts)
    parts{end + 1} = flatten(get(texts(k), "string"));
  end

  % The plotted numbers. Rewriting a line's ydata changes neither the object
  % count nor the handles, so the data has to be summarised in here too.
  kinds = {"line", "patch", "surface", "image"};
  for k = 1:numel(kinds)
    drawn = findall(handle, "type", kinds{k});
    for j = 1:numel(drawn)
      parts{end + 1} = data_summary(drawn(j));
    end
  end

  text = strjoin(parts, ";");
end


function text = data_summary(handle)
%DATA_SUMMARY Enough of an object's data to notice it changed, cheaply.
  pieces = {};
  props = {"xdata", "ydata", "zdata", "cdata"};
  for k = 1:numel(props)
    try
      values = get(handle, props{k});
    catch
      continue;   % this kind of object has no such data
    end
    if ~isnumeric(values) || isempty(values)
      continue;
    end
    values = double(values(:));
    values = values(isfinite(values));
    % A weighted sum alongside the plain one, so reordering the same numbers
    % does not go unnoticed.
    weights = (1:numel(values))';
    pieces{end + 1} = sprintf("%d:%.17g:%.17g", ...
        numel(values), sum(values), sum(values .* weights));
  end
  text = strjoin(pieces, ",");
end


function text = flatten(value)
%FLATTEN One line of text out of a char matrix, cell or string property.
  if iscell(value)
    text = strjoin(cellfun(@flatten, value, "UniformOutput", false), " ");
  elseif ischar(value)
    text = strjoin(cellstr(value)', " ");
  else
    text = "";
  end
end


function ok = publish(folder, number, handle)
%PUBLISH Write the figure and its metadata, metadata first.
%   The viewer keys off the PNG, so by the time it sees the image the sidecar it
%   reads is already in place.
  ok = false;
  if isempty(findall(handle, "type", "axes"))
    return;   % nothing drawn yet; print would refuse
  end

  try
    metadata = jsonencode(struct("number", number, ...
                                 "title", figure_title(handle, number), ...
                                 "source", "octave"));
    write_atomically(folder, number, ".json", @(target) write_text(target, metadata));
    write_atomically(folder, number, ".png", ...
                     @(target) print(handle, target, "-dpng", "-r150"));
    ok = true;
  catch err
    warn_once("computor_figures:publish", ...
              "Cannot publish figure %d: %s", number, err.message);
  end
end


function text = figure_title(handle, number)
%FIGURE_TITLE The most specific human-readable name the figure carries.
  text = strtrim(flatten(get(handle, "name")));
  if ~isempty(text)
    return;
  end

  allAxes = findall(handle, "type", "axes");
  for k = 1:numel(allAxes)
    text = strtrim(flatten(get(get(allAxes(k), "title"), "string")));
    if ~isempty(text)
      return;
    end
  end

  text = sprintf("Figure %d", number);
end


function write_atomically(folder, number, extension, write)
%WRITE_ATOMICALLY Fill a temporary file next to the target, then rename it over.
%   The viewer reacts to files appearing, so it must never get to read a
%   half-written PNG. A rename inside the folder is atomic. The temporary name
%   keeps the real extension, because print picks the format from it, and starts
%   with a dot so it is not mistaken for a figure -- readers match
%   fig-NNNNNN.png exactly.
  target = figure_path(folder, number, extension);
  temporary = fullfile(folder, sprintf(".fig-%06d.tmp%s", number, extension));

  unwind_protect
    write(temporary);
    [ok, message] = movefile(temporary, target, "f");
    if ~ok
      error("computor_figures:write", "Cannot write '%s': %s", target, message);
    end
  unwind_protect_cleanup
    if isfile(temporary)
      delete(temporary);
    end
  end_unwind_protect
end


function write_text(target, text)
%WRITE_TEXT Write UTF-8 text, so a title with an umlaut survives the trip.
  fid = fopen(target, "w");
  if fid < 0
    error("computor_figures:write", "Cannot open '%s' for writing", target);
  end
  unwind_protect
    fprintf(fid, "%s\n", text);
  unwind_protect_cleanup
    fclose(fid);
  end_unwind_protect
end


function withdraw(folder, number)
%WITHDRAW Take a figure out of the folder. Deleting the PNG is what closes it.
  delete_if_present(figure_path(folder, number, ".png"));
  delete_if_present(figure_path(folder, number, ".json"));
end


function delete_if_present(path)
  if isfile(path)
    delete(path);   % a file the viewer removed in the same instant is not an error
  end
end


function path = figure_path(folder, number, extension)
  path = fullfile(folder, sprintf("fig-%06d%s", number, extension));
end


function warn_once(identifier, template, varargin)
%WARN_ONCE Say it the first time. A failing print fails on every tick, and the
%   repetition would bury the student's own output.
  persistent seen = {};
  if any(strcmp(seen, identifier))
    return;
  end
  seen{end + 1} = identifier;
  warning(identifier, template, varargin{:});
end
